npeps/npep-133-fqdn-egress-selector.md

```diff
@@ -0,0 +1,428 @@
+  Pods must be able to talk to `my-service.com`.
+
+* As a cluster admin, I want to allow Pods in the "monitoring" namespace to be
+  able to send traffic to a logs-sink, hosted at `logs-storage.com`
+
+* As a cluster admin, I want to allow all Pods in the cluster to send traffic to
+  any of the managed services provided by my Cloud Provider. Since the cloud
+  provider has a well known parent domain, I want to allow Pods to send traffic
+  to all sub-domains using a wild-card selector -- `*.my-cloud-provider.com`
+
+### Future User Stories
+
+These are some user stories we want to keep in mind, but due to limitations of
+the existing Network Policy API, cannot be implemented currently. The design
+goal in this case is to ensure we do not make these unimplementable down the
+line.
+
+* As a cluster admin, I want to switch the default disposition of the cluster to
+  be default deny. This is enforced using a `BaselineAdminNetworkPolicy`. I also
+  want individual namespace owners to be able to specify their egress peers.
+  Namespace admins would then use a FQDN selector in the Kubernetes
+  `NetworkPolicy` objects to allow `my-service.com`.
+  
+## API
+
+This NPEP proposes adding a new type of `AdminNetworkPolicyEgressPeer` called
+`FQDNPeerSelector` which allows specifying domain names.
+
+```golang
+
+// Domain describes one or more DNS names to be used as a peer.
+//
+// Domain can be an exact match, or use the wildcard specifier '*' to match one
+// or more labels.
+//
+// '*', the wildcard specifier, matches one or more entire labels. It does not
+// support partial matches. '*' may only be specified as a prefix.
+// 
+//  Examples:
+//    - `kubernetes.io` matches only `kubernetes.io`.
+//      It does not match "www.kubernetes.io", "blog.kubernetes.io",
+//      "my-kubernetes.io", or "wikipedia.org".
+//    - `blog.kubernetes.io` matches only "blog.kubernetes.io".
+//      It does not match "www.kubernetes.io" or "kubernetes.io".
+//    - `*.kubernetes.io` matches subdomains of kubernetes.io.
+//      "www.kubernetes.io", "blog.kubernetes.io", and
+//      "latest.blog.kubernetes.io" match, however "kubernetes.io", and
+//      "wikipedia.org" do not.
+//
+// +kubebuilder:validation:Pattern=`^(\*\.)?([a-zA-z0-9]([-a-zA-Z0-9_]*[a-zA-Z0-9])?\.)*[a-zA-z0-9]([-a-zA-Z0-9_]*[a-zA-Z0-9]))?\.?$`
+type Domain string
+
+type AdminNetworkPolicyEgressPeer struct {
+    <snipped>
+    // Domains provides a way to specify domain names as peers.
+    //
+    // Domains is only supported for ALLOW rules. In order to control access,
+    // Domain ALLOW rules should be used with a lower priority egress deny --
+    // this allows the admin to maintain an explicit "allowlist" of reachable
+    // domains.
+    //
+    // Support: Extended
+    //
+    // <network-policy-api:experimental>
+    // +optional
+    // +listType=set
+    // +kubebuilder:validation:MinItems=1
+    Domains []Domain `json:"domains,omitempty"`
```

Comment:
In the meeting @npinaeva Proposed `DomainName` here however regardless Antonio suggested we always reference an RFC for decisions like this

Reply:
https://datatracker.ietf.org/doc/html/rfc1034 S3.1 is the relevant one; 

> A domain is identified by a domain name, and consists of that part of
> the domain name space that is at or below the domain name which
> specifies the domain.

So a _Domain_ is a subtree of domain space whereas a _Domain Name_ is the textual name of one particular node in that tree.  Not sure that gives a clear answer to our naming dilemma though!  `*` patterns represent subtrees, FQDNs represent single nodes in the tree! Perhaps `DomainNamePatterns` conveys exactly what our field represents (but I think `Domains` and `DomainNames` are both defensible/attackable depending on how closely you read the RFC :laughing: )?

The "." suffix is also explained in that section.  A "." at the end of the domain name means it is absolute, whereas no "." means that it is relative to the local domain and the DNS resolver should add suffixes as needed.

Reply:
I think a domain name with `*` is still considered a domain name. I read RFC as "domain name is a string, domain is an internal structure of the domain name space".
If we read closer to the wildcards definition S4.3.3, it says

> In the previous algorithm, special treatment was given to RRs with owner
names starting with the label "*".

S3.6 defines

> When we talk about a specific RR, we assume it has the following:
owner     -      which is the domain name where the RR is found.

So owner name is the domain name, and owner name may start with `*`, which makes it a valid domain name?
Reading RFCs is fun :D

Reply:
They do say that the owner is a domain(!) but they also say 
```
If any match, copy them into the answer
section, but set the owner of the RR to be QNAME, and
not the node with the "*" label.
```
I.e. don't send `*` on the wire and 
```
The owner name of the wildcard RRs is of
the form "*.<anydomain>", where <anydomain> is any domain name.
<anydomain> should not contain other * labels
```
Which implicitly says that a wildcard is not a domain so, it's a bit contradictory!

I think the wildcard comes from implementation concerns, they're really talking about the configuration for a DN server (and probably a particular one at that) and saying "you'll need wildcards in your implementation to handle mail and they should work like this to avoid these problems".